### PR TITLE
Block Library: Add width attribute for resizable Column blocks

### PR DESCRIPTION
**Pull request by aduth**
This pull request seeks to implement an initial exploration of resizable column blocks. It should serve as some foundation upon which to improve the usefulness of the Columns block.

columns

In this first iteration:

- An empty Column block displays with the "Button" appender introduced in #14241, similar to the Group block (#14943). The thinking is that this allows for a simpler gauge of columns width before adding content.
- An individual column block can be assigned with a `width` attribute which overrides the default equal distribution of column widths in a Columns block. Widths of all adjacent Column blocks are updated automatically, and widths redistributed when a Column is added or removed from the Columns block.

Not yet included, but planned for future iterations:

- A draggable resizer handle to adjust columns width. I have the bulk of the code for this complete already.
- Quick-select "template" options for columns distribution when initially creating a Columns block.
- Improved methods to select a Column and a Columns block to adjust settings.

**Specific Behavior:**

- When changing the width of a Column, the amount of increase or decrease is respectively removed or added in equal distribution across the Column blocks _after_ the block being changed, unless the Column block is the last in the set (in which case, the redistribution occurs in reverse).
- When adding or removing a Column to a Columns block, the new Column is assigned a width as if there were equal distribution. For example, increasing Columns from 2 to 3 would assume a width of 33.3% for the new block (100 / 3). This width (or, in the case of removal, the width of the removed block) is removed from or added to other blocks in the Columns.
- The default behavior of Columns remains unchanged in that `width` is only assigned or adjusted if a width value was assigned to a Column block in the set. This also helps ensure backwards-compatibility with previously-existing Column blocks.

**Testing Instructions:**

Verify that you can resize a Column block, and that in doing so, adjacent Column blocks are resized accordingly. Similarly, verify that adding or removing new Columns should redistribute widths to/from other columns.

1. Navigate to Posts > Add New
2. Insert a Columns block
3. Select the Column block (ArrowUp and ArrowDown are reliable here)
4. Adjust width value in the Inspector
5. Observe that the second column updates its width accordingly (to a sum of 100)
6. Select the Columns block
7. Increase / decrease Colujmsn count
8. Observe that the previous two columns update their widths accordingly (to a sum of 100)
9. Repeat steps 3 through 8 with a variation of Column combinations

**Remaining tasks:**

- [x] There is occasionally some issue in columns redistribution related to precision, where the sum total of Column widths may not equal 100 (but rather, e.g. 99.98%)

### CODE REVIEW

**Comment by paulwilde** [comment]
It would be ideal if the columns could snap into a grid of 12 as is common with most grid foundations such as Bootstrap, or at least if there's a preference to enable such functionality. Reason being this allows for a consistent grid on a website rather than having random widths on the page.

Having inline styles is also very troublesome when it comes to making these columns responsive, especially if the ability to set different widths on different breakpoints were to be added in the future.

File: packages/block-library/src/column/edit.js
Comment (youknowriad) [inline packages/block-library/src/column/edit.js]: I like this change :)

Comment (youknowriad) [reply packages/block-library/src/column/edit.js]: Do you why the sibling inserter is not shown when there's a single column with a single block? It feels like a separate issue but it's made more obvious with this change

File: packages/block-library/src/column/edit.js
Comment (youknowriad) [inline packages/block-library/src/column/edit.js]: Nice little perf tweak

File: packages/block-library/src/column/edit.js
Comment (youknowriad) [inline packages/block-library/src/column/edit.js]: Nit: Any reason you didn't destructure registry

File: packages/block-library/src/column/edit.js
Comment (youknowriad) [inline packages/block-library/src/column/edit.js]: I wonder if the computations here could be extracted to a unit-testable pure function.

File: packages/block-library/src/column/edit.js
Comment (youknowriad) [inline packages/block-library/src/column/edit.js]: I wonder if the computations here could be extracted to a unit-testable pure function.

File: packages/block-library/src/column/edit.js
Comment (aduth) [reply packages/block-library/src/column/edit.js]: > Nice little perf tweak

Thanks for reminding: I'd meant to reference some commentary at https://github.com/WordPress/gutenberg/pull/13899#discussion_r266877788 which had led to some of this general (but related/required) refactoring.

File: packages/block-library/src/column/save.js
Comment (youknowriad) [inline packages/block-library/src/column/save.js]: is this kses allowed (just to make sure we check as all inline style properties are not allowed)?

File: packages/block-library/src/column/edit.js
Comment (aduth) [reply packages/block-library/src/column/edit.js]: > Nit: Any reason you didn't destructure registry

I'm personally not a fan of destructuring in an arguments signature, since it loses context of _what it is_ that is being destructured. To a lesser degree, I find it hard to visually distinguish between proper arguments, and properties destructured (attention to the `{` and `}`).

File: packages/block-library/src/column/edit.js
Comment (aduth) [reply packages/block-library/src/column/edit.js]: > I wonder if the computations here could be extracted to a unit-testable pure function.

Yeah, I expect there's a fair bit of clean-up which could be done here 😅 Perhaps even some reusability between redistribution logic between the individual Column and its parent Columns block.

File: packages/block-library/src/columns/edit.js
Comment (youknowriad) [inline packages/block-library/src/columns/edit.js]: Same, I wonder if this could be unit testable

**Review by youknowriad — COMMENTED** [review]
Overall, the PR seems great to me. It would open a lot more possibilities for follow-up improvements (drag and drop, setup state)

File: packages/block-library/src/column/save.js
Comment (aduth) [reply packages/block-library/src/column/save.js]: > is this kses allowed (just to make sure we check as all inline style properties are not allowed)?

Good catch! It is not.

I will need to filter this set, and follow-up with a Trac ticket.

https://core.trac.wordpress.org/browser/trunk/src/wp-includes/kses.php?rev=45242#L2057

**Comment by youknowriad** [comment]
@Soean you're commenting from the future mate :) (Github bug)

**Comment by mapk** [comment]
I just tested today and immediately loved the new Button appenders in the columns! Fantastic indicators and visual reference for the columns.

I've got a few notes for some improvements before we merge this.

1. The column width slider defaults with no number in it. But it's set at 50%. Can we add the number there by default?
2. It would be great to add a `%` sign, or maybe just change the label to say `Percentage width` for the column width slider so that it's obvious we're setting percentage widths.
3. Can we also add a `Reset` button to return the columns back to the default setting?

Something like this:

allgood

**Comment by mapk** [comment]
Or maybe the reset button can be similar to the Color Settings for the paragraph block and say, `clear` instead? This way we follow a pattern.

Screen Shot 2019-05-08 at 4 00 31 PM

**Comment by draganescu** [comment]
Hi @aduth I've tested and found that the width in certain cases (kinda extreme) make the columns flicker:

collumns-flicker

**Comment by aduth** [comment]
@draganescu I had seen that on an occasion as well. It's odd, because the only effective difference in markup generated is the application of a `flex-basis` style, which shouldn't seem like it ought to have this effect. I'd wondered if it was perhaps a browser bug. If you are able to reproduce it, could you try viewing it in another browser?

**Comment by aduth** [comment]
Thanks for the feedback, all. I'll plan to apply it today.

**Comment by aduth** [comment]
> * The column width slider defaults with no number in it. But it's set at 50%. Can we add the number there by default?

Technically it has an undefined default width. That it appears at the halfway point is incidentally how the range input is shown when there is no value assigned. We _could_ force an an explicit width for all Column blocks, but I'm reluctant since it both seems unnecessary (equal distribution is a simple and safe default) and could have implications on backwards-compatibility of existing columns block content. We _could_ display a value representing the "effective" width (either the explicit width of the assumed width of a 100 / # columns), but this may mislead someone to think there is a width assigned when there is not.

It's similar to the paragraph block, where a numeric value is not shown in the font size field unless explicitly assigned, even if technically there is a value which could be computed.

image

**Comment by aduth** [comment]
> 2\. It would be great to add a `%` sign, or maybe just change the label to say `Percentage width` for the column width slider so that it's obvious we're setting percentage widths.

The latter is simpler, though I'd acknowledge a `%` sign could further reinforce the unit. There's some technical challenges with implementing this as `<input type="number">`, which has some obvious useful semantics and usability / accessibility perks, though also limiting as far as what's considered a valid value. In other words, it would probably not be advisable to try to include `%` as part of the value itself. The next best thing I could imagine might be similar to the idea of [Bootstrap's input gorup "add-ons"](https://getbootstrap.com/docs/4.0/components/input-group/). This would be more a visual flourish though, and a more descriptive label would be advisable in any case.

**Comment by mapk** [comment]
> It's similar to the paragraph block, where a numeric value is not shown in the font size field unless explicitly assigned, even if technically there is a value which could be computed.

Ah yeah, I see. It even feels odd for me in the font size input too. At least the font size has the "Normal" indicator next to it.

Well, let's leave it blank for now so as not to hold up this PR. Would a reset button just force it back to a blank state?

**Comment by aduth** [comment]
> 3\. Can we also add a `Reset` button to return the columns back to the default setting?

>Or maybe the reset button can be similar to the Color Settings for the paragraph block and say, clear instead? This way we follow a pattern.

@mapk Do you have a preference one way or the other? As in the screenshot of my comment at https://github.com/WordPress/gutenberg/pull/15499#issuecomment-490998165 , there's precedent for both as with the paragraph block. Visually, the input in this block appears much more similar to the paragraph font size field than it does the color picker.

That being said, the "Clear" and "Reset" actions seem to share an identical effect in both of these examples. There might be room for improvement here by consolidating the verbiage to one or the other.

**Comment by aduth** [comment]
> Well, let's leave it blank for now so as not to hold up this PR. Would a reset button just force it back to a blank state?

Yes, that would be my expectation.

File: packages/block-library/src/column/edit.js
Comment (aduth) [reply packages/block-library/src/column/edit.js]: > Do you why the sibling inserter is not shown when there's a single column with a single block? It feels like a separate issue but it's made more obvious with this change

I think it's the same as the fact that there is no sibling inserter shown at the end of the block list, only the "default appender". The column is effectively the same as the top-level block list.

However, I don't necessary agree that the sibling inserter shouldn't be shown after the last block in a block list. I think there may have been an issue about this at some point?

File: packages/block-library/src/column/save.js
Comment (aduth) [reply packages/block-library/src/column/save.js]: > is this kses allowed (just to make sure we check as all inline style properties are not allowed)?

Added in 5bb641cb1881f3d7bcea6513714166c3d570a727

File: packages/block-library/src/column/edit.js
Comment (aduth) [reply packages/block-library/src/column/edit.js]: > I wonder if the computations here could be extracted to a unit-testable pure function.

Still generally more complex than I'd have liked, but improved refactoring and tests in 6880cfa731c2968e9ca59f426e4ba1047ab52578.

**Comment by aduth** [comment]
> It would be ideal if the columns could snap into a grid of 12 as is common with most grid foundations such as Bootstrap, or at least if there's a preference to enable such functionality. Reason being this allows for a consistent grid on a website rather than having random widths on the page.
>
>It would also allow for each column to use classes instead of inline styles which are quite troublesome when it comes to making these columns responsive, especially if the ability to set different widths on different breakpoints were to be added in the future.

The styling point is compelling, and while I think a grid layout generally would prove useful especially in the long-term vision of full-site editing, it seems like a challenge with much broader impact than the columns block specifically. Ultimately I'd defer to others here on the design guidance, and I'd be wary of potential future incompatibility if such a grid system were to be implemented, but I think there's room for either both or a future enhancement toward grid layouts that this could serve as a smaller iterative improvement toward.

**Comment by aduth** [comment]
Functionally I think the branch is in a good state. There's a legitimate failing end-to-end test which still needs to be updated, due to changes in the rendered appender for the columns block (it no longer inserts paragraphs by default). I'll plan to fix this up in my morning.

**Comment by jasmussen** [comment]
Very nice work. GIF:

columbns

The primary challenge with this branch is one that is more existential to the columns block itself. There are three levels of nesting, and to resize a column you have to first select the column. I would suggest that #15537 will make this substantially easier to do, so that's good.

However it's still a user interface we should consider improving in the future. Media & Text shows that it's possible to select the parent block itself, and get access to resizing controls even for child elements:

media   text

I would hope that we can consider an interface such as the above for the columns block as well, one where the individual column blocks never need to be selected, but are instead surfaced as interactive when the _columns_ block itself is selected. The idea here is that there is little value in letting you select individual columns, and a basic way we could address this is:

- Select the Columns block and see blue resizing dots/lines between every column. I.e. 2 dots if you have three columns.
- In the sidebar when the columns block is selected, there is an input field for each column present, where you can type in the number. 
- Individual column blocks are not selectable. 

Question: does this branch preclude exploring such an approach for the Columns block in the future?

**Comment by youknowriad** [comment]
This is looking good to me, it seems though that the e2e failures are legit?

**Comment by aduth** [comment]
>Question: does this branch preclude exploring such an approach for the Columns block in the future?

To the point of a draggable interaction: No. In fact, it was initially a target for this initial branch. I'd completed 90% of the work and expected to finish the other 90% in a subsequent pull request.

To the point about managing individual column widths from the "Columns" block, rather than from the "Column" block: I don't think this branch would make that any more difficult to achieve. I still sense the width attribute would remain at the column itself. Most of what is implemented here in the Column block would necessarily have to move to the Columns block instead. It begs the question whether it ought to be done now, rather than later. Doing so would likely delay its initial introduction to the 5.8 plugin release rather than to next week's 5.7. It doesn't bother me to do so, and I'd grant that managing all of the column settings from the top-level Columns block might make for a simpler experience on the whole.

> This is looking good to me, it seems though that the e2e failures are legit?

The tests needed to be updated in consideration of the button appender. The changes in 16b3ae0 should hopefully clear it up.

**Comment by jasmussen** [comment]
Thanks for the clarification! All of what you say sounds good. It seems like the most important thing is that we don't make it harder on ourselves, markup deprecation wise, with these features, and since that doesn't seem to me the case it's not critical the column refactor ux change happen now as opposed to later.

**Comment by chrisvanpatten** [comment]
Is there any thought an option about disabling the sliders/free resizing (in the same way custom font size can be disabled)?

**Comment by youknowriad** [comment]
@chrisvanpatten I think we need to have this discussion more holistically and I'd like to see something like this explored https://github.com/WordPress/gutenberg/issues/6023#issuecomment-485372463

I agree that as we're adding more and more features for blocks, this starts to have some priority.

**Comment by aduth** [comment]
The current build failure seems intermittent, though definitely related to the columns interaction. I sense it's related to intermittent failures which already occur in `master` around these same tests (`writing-flow`, `block-hierarchy-navigation`). I suspect it might be related to asynchronous rendering behaviors of a block's own inner blocks. There's probably some improvements to be made here around the [relevant code](https://github.com/WordPress/gutenberg/blob/94d48e09dcfb547ad1b142d137a51a378d57077f/packages/block-editor/src/components/block-list/index.js#L204-L223) to consider block hierarchy, but I expect it would be much more impactful than what I'd want to address here. I'll see if I can improve the reliability of the tests with e.g. some `page.waitForSelector` to allow for the asynchronous rendering to complete (if my hypothesis is correct, that is).

**Comment by youknowriad** [comment]
@aduth I personally wonder if it's related to the change in the appender as the number of "arrow" clicks to navigate between columns may change with this change and If I remember properly this test (as I had to fix it some times), it relies on arrow navigation.

**Comment by aduth** [comment]
@youknowriad Since it's only intermittently failing (i.e. sometimes passing, at least locally), I don't think it's a fundamental issue with the behavior of the test case.

**Comment by aduth** [comment]
I spent the better part of the day trying to work with the end-to-end tests to improve their stability. I think they've always been quite unstable surrounding these specific behaviors, but the specific introduction of the button block appender made things even more difficult.

I'm not particularly happy with the inevitable solution I came up with for the tests to pass in a840fd9, but I think there are some larger problems with the tests that need a more focused effort, notably:

- Delays in [`Fill`](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/slot-fill) unmounting when a new set of `Fill`s occupies a `Slot` (when switching from one block to another)
- `BlockList` [rendering as async](https://github.com/WordPress/gutenberg/blob/4108e37d3771b5871cd720216432b5bbd5255551/packages/block-editor/src/components/block-list/index.js#L210-L213) both for:
   - Inner blocks of a block
   - A block which receives focus but has not yet become selected (i.e. WritingFlow arrow navigation, splitting, merging)

Since these impact exclusively the tests and the behavior of the columns is otherwise agreed upon, I'd personally suggest we merge this in its current form and iterate on test improvement separately.

**Review by youknowriad — APPROVED** [review]
Makes sense to me. Let's work on these tests separately.

File: packages/block-library/src/columns/style.scss
Comment (aduth) [inline packages/block-library/src/columns/style.scss]: I'm assuming this wasn't strictly necessary, also by the fact that it (wrongly) assumes a Columns block would only ever have 2 columns. In some testing, the default width distribution of a columns without widths assigned (`flex-basis: auto;`) appears to be equal distribution.

Comment (aduth) [reply packages/block-library/src/columns/style.scss]: > I'm assuming this wasn't strictly necessary, also by the fact that it (wrongly) assumes a Columns block would only ever have 2 columns. In some testing, the default width distribution of a columns without widths assigned (`flex-basis: auto;`) appears to be equal distribution.

Actually, in some final testing, it does appear to make a difference. I'm not sure if it's _intended_ that mid-range viewports show as two-column, but that's how it behaves in master, and I'd rather leave it as it was.

image

(Note: This applies only to Columns blocks where widths are not explicitly assigned)

**Comment by aduth** [comment]
> Hi @aduth I've tested and found that the width in certain cases (kinda extreme) make the columns flicker:

This seems to happen when columns contain images. In some cross-browser testing, I believe it's an issue specifically affecting Chrome. It's not a great experience surely, but based on my previous comment on how limited the DOM changes to a Column block is with this branch, I strongly suspect this is a browser rendering bug largely out of our control.

File: packages/block-library/src/columns/style.scss
Comment (m-e-h) [reply packages/block-library/src/columns/style.scss]: Your assumptions sound correct here @aduth 
The original design had 2 columns always showing at the medium breakpoint. You may not have noticed it cause It didn't actually work for the first couple versions of the editor. Got fixed here https://github.com/WordPress/gutenberg/pull/12199

I don't think that two column breakpoint would be missed too much if removing it would help what you're doing here.

**Comment by paaljoachim** [comment]
This feels like one of multiple very useful steps/iterations that will affect width, height, padding, margin, drag, responsive, grid and more of multiple blocks (perhaps all blocks). It would be great with a writeup of this step and thoughts on future steps. So that others could also share feedback. Thanks!

**Comment by mapk** [comment]
🎉  Excellent work! 

Will there be another PR soon that works toward the dragging resizing option that Joen [mentioned above](https://github.com/WordPress/gutenberg/pull/15499#issuecomment-491194849)?

**Comment by aduth** [comment]
> Will there be another PR soon that works toward the dragging resizing option that Joen [mentioned above](https://github.com/WordPress/gutenberg/pull/15499#issuecomment-491194849)?

I'll create some follow-up issues today.

**Comment by aduth** [comment]
Follow-up issues:

- [Block Library: Column: Add draggable resize control](https://github.com/WordPress/gutenberg/issues/15659) (#15659)
- [Block Library: Column: Move resizing controls to Columns block](https://github.com/WordPress/gutenberg/issues/15660) (#15660)
- [Block Editor: Add predefined arrangements options support to InnerBlocks](https://github.com/WordPress/gutenberg/issues/15662) (#15662)
- [Block Library: Column: Add predefined "layout" selections for new Columns block](https://github.com/WordPress/gutenberg/issues/15663) (#15663)
- [Editor: KSES: Include Flex properties in CSS whitelist to support core Column block width](https://core.trac.wordpress.org/ticket/47281) ([Trac#47281](https://core.trac.wordpress.org/ticket/47281))

**Comment by StaggerLeee** [comment]
There are already Bootstrap Gutenberg blocks that you will never come close to. And you lose your time to make everything from the scratch, when there is almost perfect Bootstrap library to integrate.
I do not understand it.

Does Matt Mullenweg pay you per code lines ?

Ah OK, now continue to adapt them to responsive views.

**Comment by SchneiderSam** [comment]
If I enter 33 in one of my columns, the entire column layout will be destroyed. Since the page is online, I can't show it live. But somewhere there are still problems.

Login: rb
PW: buvexayu
URL: http://b320kx.myraidbox.de/

Normally it should look like this: 
Screenshot_3


but in the live site it looks like this:
Screenshot_4

**Comment by aduth** [comment]
@StaggerLeee That's an awfully cynical stance to hold, and I think you could express your point without resorting to sarcasm or abuse. It's fine for a plugin and core functionality to coexist in the same space. The [WordPress philosophy](https://wordpress.org/about/philosophy/) advocates design for the majority, simplicity, and out-of-the-box usefulness. A plugin can add specialized functionality, and simply absorbing one plugin's implementation isn't always the best option for the baseline offering; though it can certainly _inform_ it, such as is already taking place in follow-up tasks (e.g. #15662 "Prior art"). If you have specific feedback, I'd encourage you to provide it (constructively) in one of the existing issues, a new issue, or to participate in one of the [weekly meetings](https://make.wordpress.org/meetings/).

@SchneiderSam Could you create a [new issue](https://github.com/WordPress/gutenberg/issues/new?template=Bug_report.md) for this? It may be a specific incompatibility with your theme, but a dedicated issue would be the more appropriate place for debugging.

**Comment by nerrad** [comment]
@StaggerLeee please take some time to read the [Code of Conduct](https://github.com/WordPress/gutenberg/blob/master/CODE_OF_CONDUCT.md) for this repository.  Your comments could be seen as breaching some of the principles of that code.  We always welcome constructive feedback here but repeated abuse of the code of conduct will not be tolerated.

**Comment by karmatosed** [comment]
@StaggerLeee whilst I understand you may have a strong opinion here, @nerrad is correct there is a code of conduct that applies to this repo. I would ask you to please be mindful of this in the future. Every person here is doing their best and individual humans that deserve respectful interactions. Let's focus on all creating the best product we can together.

Thank you @nerrad for your response on this issue 🙏

**Comment by StaggerLeee** [comment]
What a fuss, just for mentioning the Boss. My apologies.

**Comment by karmatosed** [comment]
Whilst again I understand your feelings here, the code of conduct is taken seriously within this repo and WordPress itself. Let's draw a line for now under this and move on with that clear understanding for future interactions.

**Comment by m-e-h** [comment]
@StaggerLeee have you checked out the Joomla CMS? Pretty sure it still includes the bootstrap framework in it's core. Of course there are [challenges](https://joomla.digital-peak.com/blog/205-the-joomla-bootstrap-problem) to such an approach but it sounds like it may be closer to what you're looking for.

I don't get paid anything for my code here :man_shrugging:

**Pull request merged**
